src/storage-controller/src/collection_mgmt.rs

```diff
@@ -192,13 +340,477 @@ where
+            Ok(Err(StorageError::InvalidUppers(failed_ids))) => {
+                // Someone else wrote to the collection or bumped the upper. We
+                // need to sync to latest persist state and potentially patch up
+                // our `to_write`, based on what we learn and `desired`.
+
+                assert_eq!(
+                    failed_ids.len(),
+                    1,
+                    "received errors for more than one collection"
+                );
+                assert_eq!(
+                    failed_ids[0].id, self.id,
+                    "received errors for a different collection"
+                );
+
+                let actual_upper = if let Some(ts) = failed_ids[0].current_upper.as_option() {
+                    ts.clone()
+                } else {
+                    return ControlFlow::Break("upper is the empty antichain".to_string());
+                };
+
+                tracing::info!(%self.id, ?actual_upper, expected_upper = ?self.current_upper, "upper mismatch while bumping upper, syncing to persist state");
+
+                self.current_upper = actual_upper;
+
+                self.sync_to_persist().await;
+            }
+            Ok(Err(err)) => {
+                panic!(
+                    "unexpected error while trying to bump upper of {}: {:?}",
+                    self.id, err
+                );
+            }
+            // Sender hung up, this seems fine and can happen when shutting down.
+            Err(_recv_error) => {
+                // Exit the run loop because there is no other work we can do.
+                return ControlFlow::Break("persist worker is gone".to_string());
+            }
+        }
+
+        ControlFlow::Continue(())
+    }
+
+    fn handle_shutdown(&mut self) {
+        let mut senders = Vec::new();
+
+        // Prevent new messages from being sent.
+        self.cmd_rx.close();
+
+        // Get as many waiting senders as possible.
+        while let Ok((_batch, sender)) = self.cmd_rx.try_recv() {
+            senders.push(sender);
+
+            // Note: because we're shutting down the sending side of `rx` is no
+            // longer accessible, and thus we should no longer receive new
+            // requests. We add this check just as an extra guard.
+            if senders.len() > CHANNEL_CAPACITY {
+                // There's not a correctness issue if we receive new requests,
+                // just unexpected behavior.
+                tracing::error!("Write task channel should not be receiving new requests");
+                break;
+            }
+        }
+
+        // Notify them that this collection is closed.
+        //
+        // Note: if a task is shutting down, that indicates the source has been
+        // dropped, at which point the identifier is invalid. Returning this
+        // error provides a better user experience.
+        notify_listeners(senders, || Err(StorageError::IdentifierInvalid(self.id)));
+    }
```

Comment:
Yes, I should have expected that! This code path is only taken when removing a collection that is manged by this. But this code is only used for builtin storage-managed collections (the introspection collections), and we normally never remove them, or can't even do that.

We _might_ add code in the future that does drop these collections, in which case this code path will come in handy, but it might rot! Would you prefer I remove this, and add something akin to `unreachable!()`, or leave it in?

Reply:
Something like `unreachable!()` sounds good